plugin/src/main/kotlin/tanvd/grazi/ide/language/markdown/MarkdownSupport.kt

```diff
@@ -2,73 +2,53 @@ package tanvd.grazi.ide.language.markdown
 
 
 import com.intellij.psi.PsiElement
-import org.intellij.plugins.markdown.lang.MarkdownElementTypes
-import org.intellij.plugins.markdown.lang.MarkdownTokenTypes
-import tanvd.grazi.grammar.*
+import com.intellij.psi.util.PsiTreeUtil
+import tanvd.grazi.grammar.GrammarChecker
+import tanvd.grazi.grammar.Typo
 import tanvd.grazi.ide.language.LanguageSupport
-import tanvd.grazi.utils.spellcheckOnly
-import tanvd.grazi.utils.traverse
+import tanvd.grazi.utils.parents
 
 class MarkdownSupport : LanguageSupport() {
     companion object {
         private val bulletsIgnoredCategories = listOf(Typo.Category.CASING)
     }
 
     override fun isRelevant(element: PsiElement): Boolean {
-        return MarkdownPsiUtils.isHeader(element) || MarkdownPsiUtils.isParagraph(element)
-                || MarkdownPsiUtils.isCode(element) || MarkdownPsiUtils.isOuterListItem(element)
+        return MarkdownPsiUtils.isHeader(element) || MarkdownPsiUtils.isParagraph(element) || MarkdownPsiUtils.isCode(element)
     }
 
     override fun check(element: PsiElement): Set<Typo> {
-        return when {
-            MarkdownPsiUtils.isHeader(element) -> {
-                val ignoreFilter = TokensFilter()
-                val elements = element.filterForTextTokensExcluding(*MarkdownPsiUtils.inlineTypes.toTypedArray())
-
-                ignoreFilter.populateMd(elements)
-
-                ignoreFilter.filter(GrammarChecker.default.check(elements))
-
+        return GrammarChecker.default.check(listOf(element), indexBasedIgnore = { token, index ->
+            val elementInToken = token.findElementAt(index)!!
+            !MarkdownPsiUtils.isText(elementInToken) || elementInToken.parents().any { MarkdownPsiUtils.isInline(it) }
+        }).filter { typo ->
+            val typoStartElement: PsiElement = element.findElementAt(typo.location.range.start)!!
+            val typoEndElement: PsiElement = element.findElementAt(typo.location.range.last)!!
+
+            // Inline elements inside typo
+            if (generateSequence(typoStartElement) { PsiTreeUtil.nextLeaf(it) }.takeWhile { it != typoEndElement }
+                            .flatMap { it.parents() }.any { MarkdownPsiUtils.isInline(it) }) return@filter false
+
+
+            // Inline element right before a typo (TODO add categories filter)
```

Comment:
Why it is so complex? :) Can not it be replaced with getting `previousLeaf` of start element of Typo and checking if it is inline element or one of its parents is inline element?

Reply:
Unfortunately there is a probability that next/prev leaf will be a whitespace symbol and leaf after that will be inline